plume-common/src/activity_pub/inbox.rs

```diff
@@ -1,48 +1,609 @@
+                    }
+                };
 
-    fn notify(&self, conn: &C) -> Result<(), Self::Error>;
+                // Same logic for "object"
+                let obj_id = match get_id(act["object"].clone()) {
+                    Some(x) => x,
+                    None => return Inbox::NotHandled(ctx, act, InboxError::InvalidObject(None)),
+                };
+                let obj = match M::from_id(
+                    ctx,
+                    &obj_id,
+                    serde_json::from_value(act["object"].clone()).ok(),
+                ) {
+                    Ok(o) => o,
+                    Err((json, e)) => {
+                        if let Some(json) = json {
+                            act["object"] = json;
+                        }
+                        return Inbox::NotHandled(ctx, act, InboxError::InvalidObject(Some(e)));
+                    }
+                };
+
+                // Handle the activity
+                match obj.activity(ctx, actor, &act_id) {
+                    Ok(res) => Inbox::Handled(res.into()),
+                    Err(e) => Inbox::Failed(e),
+                }
+            } else {
+                // If the Activity type is not matching the expected one for
+                // this handler, try with the next one.
+                Inbox::NotHandled(ctx, act, e)
+            }
+        } else {
+            self
+        }
+    }
+
+    /// Transforms the inbox in a `Result`
+    pub fn done(self) -> Result<R, E> {
+        match self {
+            Inbox::Handled(res) => Ok(res),
+            Inbox::NotHandled(_, _, err) => Err(E::from(err)),
+            Inbox::Failed(err) => Err(err),
+        }
+    }
 }
 
-pub trait Deletable<C, A> {
-    type Error;
+/// Get the ActivityPub ID of a JSON value.
+///
+/// If the value is a string, its value is returned.
+/// Otherwise we return its `id` field, if it is a string
+///
+/// # Panics
+///
+/// This function panics if the value is neither a string nor an object with an
+/// `id` field that is a string.
+fn get_id(json: serde_json::Value) -> Option<String> {
+    match json {
+        serde_json::Value::String(s) => Some(s),
+        serde_json::Value::Object(map) => map.get("id")?.as_str().map(ToString::to_string),
+        _ => None,
+    }
+}
+
+/// A trait for ActivityPub objects that can be retrieved or constructed from ID.
+///
+/// The two functions to implement are `from_activity` to create (and save) a new object
+/// of this type from its AP representation, and `from_db` to try to find it in the database
+/// using its ID.
+///
+/// When dealing with the "object" field of incoming activities, `Inbox` will try to see if it is
+/// a full object, and if so, save it with `from_activity`. If it is only an ID, it will try to find
+/// it in the database with `from_db`, and otherwise dereference (fetch) the full object and parse it
+/// with `from_activity`.
+pub trait FromId<C>: Sized {
```

Comment:
Isn't Sized implied by default?

Reply:
IIRC, the compiler threw errors when it wasn't here. :woman_shrugging:

Reply:
I'll try to remove it just to be sure.

Reply:
It was explicitly needed here, to be able to have `Result<Self, _>` here, but was not actually required on the two other types.

Reply:
ok, I did not know